Library/Homebrew/utils/shell.rb

```diff
@@ -0,0 +1,64 @@
+require 'shellwords'
+
+module Utils
+  module Shell
+    # take a path and heuristically convert it
+    # to a shell, return nil if there's no match
+    def self.path_to_shell(path)
+      # remove leading - from login shell
+      shell_name = File.basename(path).strip_prefix("-")
+      case shell_name
+      when "sh", "bash", "zsh", "csh", "tcsh", "fish", "ksh"
+        shell_name.to_sym()
+      else nil
+      end
+    end
+
+    def self.preferred_shell
+      path_to_shell(ENV.fetch("SHELL", ""))
+    end
+
+    def self.parent_shell
+      path_to_shell(`ps -p #{Process.ppid} -o command=`.chomp)
+    end
+
+    def self.csh_quote(str)
+      # shellquote breaks on csh
+      # put everything in single quotes except
+      # single quotes themselves, which are double quoted
+      # we need to prepend a \ to literal newlines and
+      # ! (history substitution character)
+      substrings = str.split("'", -1)
+      substrings.map { |s| "'" + (s.gsub(/([\n!])/) { '\\' + $1 }) + "'" }.join("\"'\"")
```

Comment:
You can make `s.gsub(/([\n!])/) { '\\' + $1 }` more efficient if you provide it directly with a replacement string instead of passing a block:

```ruby
s.gsub(/[\n!]/, "\\\\\\0")
```

(The many backslashes look somewhat horrible, but they are required because one level of backlash escapes is needed for the string and another level is needed for the replacement placeholders `\<n>`.)